sound/soc/sof/debug.c

```diff
@@ -17,6 +17,142 @@
+	s64 ipc_response_time;
+	s64 avg_response_time = 0;
+	s64 min_response_time = LONG_MAX;
+	s64 max_response_time = 0;
+	int i = 0;
+	int err;
+	unsigned long ipc_count = 0;
+	unsigned long ipc_duration_ms = 0;
+	bool flood_duration_test = false;
+#endif
+	string = kzalloc(count, GFP_KERNEL);
+	if (!string)
+		return -ENOMEM;
+
+	size = simple_write_to_buffer(string, count, ppos, buffer, count);
+	ret = size;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_IPC_FLOOD_TEST)
+	if (!strcmp(dfse->dfsentry->d_name.name, "ipc_flood_count"))
+		goto test;
+
+	if (!strcmp(dfse->dfsentry->d_name.name, "ipc_flood_duration_ms")) {
+		flood_duration_test = true;
+		goto test;
+	}
+
+	/*
+	 * write op is only supported for ipc_flood_count or
+	 * ipc_flood_duration_ms debugfs entries atm.
+	 * ipc_flood_count floods the DSP with the number of IPC's specified.
+	 * ipc_duration_ms test floods the DSP for the time specified
+	 * in the debugfs entry.
+	 */
+	return -EINVAL;
+test:
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0) {
+		dev_err_ratelimited(sdev->dev,
+				    "error: debugfs write failed to resume %d\n",
+				    ret);
+		pm_runtime_put_noidle(sdev->dev);
+		return ret;
+	}
+
+	/* set test completion criteria */
+	if (flood_duration_test)
+		ret = kstrtoul(string, 0, &ipc_duration_ms);
+	else
+		ret = kstrtoul(string, 0, &ipc_count);
+	if (ret < 0)
+		return ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * 1000000;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &reply, sizeof(reply));
+		end = ktime_get();
+
+		if (ret < 0)
+			break;
+
+		/* compute min and max response times */
+		ipc_response_time = ktime_to_ns(ktime_sub(end, start));
+		min_response_time = min(min_response_time, ipc_response_time);
+		max_response_time = max(max_response_time, ipc_response_time);
+
+		/* calculate the moving average */
+		avg_response_time +=
+			(ipc_response_time - avg_response_time) / (i + 1);
+		i++;
+
+		/* test complete? */
+		if (flood_duration_test) {
+			if (ktime_to_ns(end) >= test_end)
+				break;
+		} else {
+			if (i == ipc_count)
+				break;
+		}
+	}
+
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: ipc flood test failed at %d%s iteration\n", i,
+			(i % 10 == 1) ? "st" : "th");
+		ret = -EINVAL;
+	} else {
+		ret = size;
+	}
+
+	if (flood_duration_test)
+		dev_dbg(sdev->dev, "IPC Flood test duration: %lldms\n",
+			(long long)ipc_duration_ms);
+
+	dev_dbg(sdev->dev,
+		"IPC Flood test count: %d, Avg response time: %lldns\n",
+		i, (long long)avg_response_time);
+	dev_dbg(sdev->dev, "IPC Flood test Max response time: %lldns\n",
+		(long long)max_response_time);
+	dev_dbg(sdev->dev, "IPC Flood test Min response time: %lldns\n",
+		(long long)min_response_time);
```

Comment:
So the above could be returned via read() as ascii to user-space. You could e.g. read out results of the last test.

Reply:
@kv2019i ack. Thanks for the suggestion. Will update soon